src/components/ZoomRect/index.js

```diff
@@ -0,0 +1,342 @@
+    if (!isEqual(prevZoomAxes, currZoomAxes)) {
+      this.syncZoomingState();
+    }
+  }
+
+  onTouchStart = () => {
+    const {
+      event: { touches },
+    } = d3;
+    if (touches.length === 1) {
+      const [touch] = touches;
+      const { pageX: x, pageY: y } = touch;
+      this.lastTouch = {
+        [Axes.time]: x,
+        [Axes.x]: x,
+        [Axes.y]: y,
+      };
+    } else if (touches.length === 2) {
+      const [touchOne, touchTwo] = touches;
+      this.lastDeltas = {
+        [Axes.time]: Math.abs(touchOne.pageX - touchTwo.pageX),
+        [Axes.x]: Math.abs(touchOne.pageX - touchTwo.pageX),
+        [Axes.y]: Math.abs(touchOne.pageY - touchTwo.pageY),
+      };
+    }
+  };
+
+  // TODO: A lot of this is duplicated with `zoomed` -- maybe they can be
+  // consolidated?
+  onTouchMove = () => {
+    const { width, height } = this.props;
+
+    const distances = {
+      [Axes.time]: width,
+      [Axes.x]: width,
+      [Axes.y]: -height,
+    };
+
+    const {
+      event: { touches },
+    } = d3;
+    let updates = null;
+    if (touches.length === 1) {
+      // If there was only one touch, then it was a drag event.
+      updates = this.performTouchDrag(touches, distances);
+    } else if (touches.length === 2) {
+      // If there were two, then it is a zoom event.
+      updates = this.performTouchZoom(touches, distances);
+    } else {
+      // We don't support more complicated gestures.
+    }
+    if (updates) {
+      this.props.updateDomains(updates);
+    }
+  };
+
+  onTouchEnd = () => {
+    const {
+      event: { touches },
+    } = d3;
+    if (touches.length === 0) {
+      this.lastTouch = null;
+    } else if (touches.length === 1) {
+      this.lastDeltas = null;
+      const [touch] = touches;
+      const { pageX: x, pageY: y } = touch;
+      this.lastTouch = {
+        [Axes.time]: x,
+        [Axes.x]: x,
+        [Axes.y]: y,
+      };
+    }
+  };
+
+  performTouchDrag = (touches, distances) => {
+    const { itemIds, subDomainsByItemId, zoomAxes } = this.props;
+    const [touch] = touches;
+    const newTouchPosition = {
+      [Axes.time]: touch.pageX,
+      [Axes.x]: touch.pageX,
+      [Axes.y]: touch.pageY,
+    };
+    const updates = {};
+    itemIds.forEach(itemId => {
+      updates[itemId] = {};
+      Axes.ALL.filter(axis => zoomAxes[axis]).forEach(axis => {
+        const subDomain = (subDomainsByItemId[itemId] || {})[axis];
+        const subDomainRange = subDomain[1] - subDomain[0];
+        let newSubDomain = null;
+        const percentMovement =
+          subDomainRange *
+          ((this.lastTouch[axis] - newTouchPosition[axis]) / distances[axis]);
+        newSubDomain = subDomain.map(bound => bound + percentMovement);
+        if (newSubDomain) {
+          updates[itemId][axis] = newSubDomain;
+        }
+      });
+    });
+    this.lastTouch = { ...newTouchPosition };
+    return updates;
+  };
+
+  performTouchZoom = (touches, distances) => {
```

Comment:
I'm a bit afraid of reimplementing this functionality of d3

Reply:
I didn't want to do it, but there's basically no way around it as far as I can see. I fought d3 for a long time to get touch + mouse zoom working together, but they seem to only operate using transformations, which we basically can't use.